### PR TITLE
cross compile arm64 / amd64 images

### PR DESCRIPTION
**Pull request by thebearingedge**
## Please make sure to check the following tasks before opening and submitting a PR

* [x] I understand and have followed the [contribution guide](https://github.com/revoltchat/revolt/discussions/282)
* [x] I have tested my changes locally and they are working as intended
* [x] These changes do not have any notable side effects on other Revolt projects

---

Hello! I've got a preliminary cross-compilation (`amd64`/`arm64`) setup going with the `base`, `bonfire`, and `delta` container images here. Testing has been a bit wonky, but all six images are building on my local M1 machine.

With cross compilation, `buildx` of `base` for both platforms takes about **10 minutes** for me vs **???** with emulation 😄.

I plan to create a separate branch and try triggering some build and push Actions on my fork, but wanted to get the work out there sooner for comment in case my removal of `cargo install` in the separate service images was going to cause problems.

✌️

### CODE REVIEW

**Comment by thebearingedge** [comment]
I have read the CLA Document and I hereby sign the CLA

File: Dockerfile [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
+    libssl-dev:"${TARGETARCH}"
+COPY scripts/build-image-layer.sh /tmp/
+RUN sh /tmp/build-image-layer.sh tools
 
-# Build all crates
+# Build all dependencies
 COPY Cargo.toml Cargo.lock ./
+COPY crates/bonfire/Cargo.toml ./crates/bonfire/
+COPY crates/delta/Cargo.toml ./crates/delta/
+COPY crates/quark/Cargo.toml ./crates/quark/
```
Comment (thebearingedge) [inline Dockerfile:22]: This should let us skip re-compilation of fetched dependencies between builds.

**Comment by thebearingedge** [comment]
After experimenting with concurrent vs. sequential builds, the caching seemed to work out better with concurrent multi-arch builds.

Total build and publish time for complete cache miss is longer, but only around 45m instead of multiple hours due to building arm64 in emulation.

Happy to make more changes. If the approach is desirable, I'm also happy to make updates to `autumn` and `january`.

**Comment by insertish** [comment]
Sorry for the late response, this came in during my semester so I missed it.

> Total build and publish time for complete cache miss is longer, but only around 45m instead of multiple hours due to building arm64 in emulation.

Honestly as long as its not obscenely long (> 1 hour), then it's fine. I currently use the CI as part of the production deployment so I'd get a bit impatient having to wait several hours or otherwise having to build it myself.

> Happy to make more changes. If the approach is desirable, I'm also happy to make updates to `autumn` and `january`.

I read through the changes twice and it looks pretty good.
In terms of autumn and january, the plan is to get those merged into this repository at some point but I haven't gotten around to it.

I've just approved the action run, so if it passes again now I'll probably be good to merge.

**Review by insertish — APPROVED** [review]

**Pull request merged**